test/cli/pin.js

```diff
@@ -0,0 +1,148 @@
+/* eslint-env mocha */
+'use strict'
+
+const expect = require('chai').expect
+const series = require('async/series')
+const runOnAndOff = require('../utils/on-and-off')
+const _ = require('lodash')
+
+// use a tree of ipfs objects for recursive tests:
+//  root
+//   |`leaf
+//    `branch
+//      `subLeaf
+
+const filenames = [ 'root', 'leaf', 'branch', 'subLeaf' ]
+
+const keys = {
+  root: 'QmWQwS2Xh1SFGMPzUVYQ52b7RC7fTfiaPHm3ZyTRZuHmer',
+  leaf: 'QmaZoTQ6wFe7EtvaePBUeXavfeRqCAq3RUMomFxBpZLrLA',
+  branch: 'QmNxjjP7dtx6pzxWGBRCrgmjX3JqKL7uF2Kjx7ExiZDbSB',
+  subLeaf: 'QmUzzznkyQL7FjjBztG3D1tTjBuxeArLceDZnuSowUggXL'
+}
+
+describe('pin', () => runOnAndOff((thing) => {
+  const filesDir = 'test/test-data/tree/'
+
+  let ipfs
+
+  before((done) => {
+    ipfs = thing.ipfs
+
+    ipfs(`object put ${filesDir + 'root.json'}`)
+      .then(() => ipfs(`object put ${filesDir + 'root.json'}`))
+      .then(() => ipfs(`object put ${filesDir + 'leaf.json'}`))
+      .then(() => ipfs(`object put ${filesDir + 'branch.json'}`))
+      .then(() => ipfs(`object put ${filesDir + 'subLeaf.json'}`))
+      .then(() => {
+        done()
+      })
+  })
+
+  it('add (recursively by default)', (done) => {
+    console.log('cli pin add starting')
+    ipfs(`pin add ${keys.root}`).then((out) => {
+      console.log('cli pin add callback')
+      console.log(out)
+//      expect(out).to.eql(`pinned ${keys.root} recursively\n`)
```

Comment:
I added timestamps to these logs to make sure the sequence was clear. Here's what I'm seeing
```
   42.851 : pin add test starts (test/cli/pin.js)
   43.875 : ipfs.add called with callback (cli/commands/pin/add.js)
  *43.877 : cleanup called, which closes repo (cli/bin.js)
   44.008 : repo is closed when accessing datastore (core/components/pin.js)
   44.008 : ipfs.add callback ran (cli/commands/pin/add.js)
   44.015 : test callback ran (test/cli/pin.js)
```
It still looks to me like the timing/callbacks themselves are working within the tests and the pin code, but cleanup being called immediately rather than waiting for the callbacks finish. 

I'll focus on making the suggested changes first and then see if this is still a problem at that point.